File: 107-level-order-bottom/solution.py
from typing import Optional
from typing import List
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        result: List[List[int]] = []
        if root is None:
            return result
        dq: deque[TreeNode] = deque()
        dq.append(root)

        while len(dq) != 0:
            level: List[int] = []
            for i in range(len(dq)):
                root = dq.popleft()
                level.append(root.val)

                if root.left is not None:
                    dq.append(root.left)
                if root.right is not None:
                    dq.append(root.right)

            result.insert(0, level)
        return result
    
"""
內層的迴圈執行次數是len(dq), 但在此迴圈裡面又會去改變dq的長度, 這樣會不會導致迴圈執行不完?

答案：不會
在 Python 中，range(len(dq)) 是在 for 迴圈開始時就被計算一次，之後迴圈會依照當時固定下來的 range 物件來迭代，不會再去重新查詢 len(dq)。

直觀理解
你可以把它想成「快照」的概念：
本層開始 → 快照：len(dq) = 3 → 只處理這 3 個節點
           ↓
迴圈中 → append 子節點 → 這些子節點屬於「下一層」
           ↓
本層結束 → 下一層 while 迴圈重新計算 len(dq) → 新的快照

若你這樣寫，就會有問題（死迴圈）：
while len(dq) > 0:       # 每次都會重新檢查
    root = dq.popleft()
    dq.append(...)       # 長度一直增加，永遠跑不完
"""

File: 107-level-order-bottom/solution.go
package main

import (
	"fmt"
	"gopractice/datastructure/Tree"
	"slices"
)
type TreeNode = Tree.TreeNode
func levelOrderBottom(root *TreeNode) (result [][]int) {
	if root == nil {
		return result
	}
	deque := []*TreeNode{}
	deque = append(deque, root)

	for len(deque) > 0 {
		level := []int{}
		levelSize := len(deque)
		for i := 0; i < levelSize; i++ {
			curNode := deque[0]
			deque = deque[1:]
			level = append(level, curNode.Val.(int))
			if curNode.Left != nil {
				deque = append(deque, curNode.Left)
			}
			if curNode.Right != nil {
				deque = append(deque, curNode.Right)
			}
		}
		result = append(result, level)
	}
	slices.Reverse(result)
	return
}
func main(){
	levelOrder := []any{3, 9, 20, nil, nil, 15, 7}
	root := Tree.BuildLevelOrderTree_v2(levelOrder)
	fmt.Println("Ans = ", levelOrderBottom(root))
}

/*
Go 的 for 迴圈條件會每次重新計算

i < len(deque) 是每一次迭代都會重新計算的。當你在迴圈內執行：
deque = deque[1:]              // 長度 -1
deque = append(deque, ...)     // 長度 +1 或 +2
len(deque) 會即時變化，導致這個迴圈永遠無法正常結束（取出 1 個、加入 1~2 個，長度不會歸零）。

所以需要用 levelSize 把 deque 當下的長度先存起來, 在拿它來做為內層迴圈的執行次數
*/